crypto/bp/bp_g2.c

```diff
@@ -0,0 +1,1212 @@
+     */
+    if (G2_ELEM_is_on_curve(group, point, ctx) <= 0)
+        goto err;
+
+    ret = 1;
+
+ err:
+    BN_CTX_end(ctx);
+    BN_CTX_free(new_ctx);
+    return ret;
+}
+
+int G2_ELEM_add(const BP_GROUP *group, G2_ELEM *r, const G2_ELEM *a,
+                const G2_ELEM *b, BN_CTX *ctx)
+{
+    BN_CTX *new_ctx = NULL;
+    FP2 *t0, *t1 = NULL, *t2 = NULL, *t3 = NULL, *t4 = NULL, *t5 = NULL,
+        *t6 = NULL;
+    int ret = 0;
+
+    if (a == b)
+        return G2_ELEM_dbl(group, r, a, ctx);
+    if (G2_ELEM_is_at_infinity(group, a))
+        return G2_ELEM_copy(r, b);
+    if (G2_ELEM_is_at_infinity(group, b))
+        return G2_ELEM_copy(r, a);
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_new()) == NULL)
+        return 0;
+
+    BN_CTX_start(ctx);
+    if ((t0 = FP2_new()) == NULL || (t1 = FP2_new()) == NULL
+        || (t2 = FP2_new()) == NULL || (t3 = FP2_new()) == NULL
+        || (t4 = FP2_new()) == NULL || (t5 = FP2_new()) == NULL
+        || (t6 = FP2_new()) == NULL)
+        goto err;
+
+    /*
+     * Note that in this function we must not read components of 'a' or 'b'
+     * once we have written the corresponding components of 'r'. ('r' might
+     * be one of 'a' or 'b'.)
+     */
+
+    /*
+     * t1, t2
+     */
+    if (b->Z_is_one) {
+        if (!FP2_copy(t1, a->X))
+            goto err;
+        if (!FP2_copy(t2, a->Y))
+            goto err;
+        /*
+         * t1 = X_a
+         */
+        /*
+         * t2 = Y_a
+         */
+    } else {
+        if (!FP2_sqr(group, t0, b->Z, ctx))
+            goto err;
+        if (!FP2_mul(group, t1, a->X, t0, ctx))
+            goto err;
+        /*
+         * t1 = X_a * Z_b^2
+         */
+
+        if (!FP2_mul(group, t0, t0, b->Z, ctx))
+            goto err;
+        if (!FP2_mul(group, t2, a->Y, t0, ctx))
+            goto err;
+        /*
+         * t2 = Y_a * Z_b^3
+         */
+    }
+
+    /*
+     * t3, t4
+     */
+    if (a->Z_is_one) {
+        if (!FP2_copy(t3, b->X))
+            goto err;
+        if (!FP2_copy(t4, b->Y))
+            goto err;
+        /*
+         * t3 = X_b
+         */
+        /*
+         * t4 = Y_b
+         */
+    } else {
+        if (!FP2_sqr(group, t0, a->Z, ctx))
+            goto err;
+        if (!FP2_mul(group, t3, b->X, t0, ctx))
+            goto err;
+        /*
+         * t3 = X_b * Z_a^2
+         */
+
+        if (!FP2_mul(group, t0, t0, a->Z, ctx))
+            goto err;
+        if (!FP2_mul(group, t4, b->Y, t0, ctx))
+            goto err;
+        /*
+         * t4 = Y_b * Z_a^3
+         */
+    }
+
+    /*
+     * t5, t6
+     */
+    if (!FP2_sub(group, t5, t1, t3))
+        goto err;
+    if (!FP2_sub(group, t6, t2, t4))
+        goto err;
+    /*
+     * t5 = t1 - t3
+     */
+    /*
+     * t6 = t2 - t4
+     */
+
+    if (FP2_is_zero(t5)) {
+        if (FP2_is_zero(t6)) {
+            /*
+             * a is the same point as b
+             */
+            BN_CTX_end(ctx);
```

Comment:
Removed CTX handling from functions that do not allocate BIGNUMs directly. Notice that `ecp_smpl.c` does the same thing inside `ec_GFp_simple_add` to handle point doublings.